### PR TITLE
KEYCLOAK-10020 Add ability to request ticket permissions by name

### PR DESCRIPTION
**Pull request by GeorgeBekh**
[KEYCLOAK-10020: Unable to get RPT by the name of a resource being shared via UMA](https://issues.jboss.org/browse/KEYCLOAK-10020)

This PR adds ability to request by name permissions for shared resources

Example of a request that wouldn't work otherwise (if another user has shared resource with the requester)
```
    curl "$PROTOCOL://$HOST:$PORT/auth/realms/$REALM/protocol/openid-connect/token" \
    -H "Authorization: Bearer $ACCESS_TOKEN" \
    --data "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket&permission=$RESOURCE_NAME&audience=$RESOURCE_CLIENT_ID"
```

### CODE REVIEW

**Comment by GeorgeBekh** [comment]
If you could help me find the right test files to put tests in i would happily do so

**Comment by GeorgeBekh** [comment]
@keycloak-ci-bot test

**Review by pedroigor — CHANGES_REQUESTED** [review]
Here is the test class as well as a method that you can use as an example for your test: https://github.com/keycloak/keycloak/blob/8b6979ac18b19eb47e618e05ccffdde0c9adb6ff/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/PermissionManagementTest.java#L69.

**Comment by GeorgeBekh** [comment]
@pedroigor thanks for the tip, but this functionality was added to support new way of obtaining an RPT, not permission ticket.

I see there are tests which obtain RPT: 
https://github.com/keycloak/keycloak/blob/6aa90963614d607e1bc4ba83300007bf2d96d9af/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/UserManagedAccessTest.java#L231

In these tests RPTs are obtained via permission ticket, but to test my PR i need to obtain RPT without permission ticket

From the keycloak doc:
> 
> permission
> 
> This parameter is optional. A string representing a set of one or more resources and scopes the client is seeking access. This parameter can be defined multiple times in order to request permission for multiple resource and scopes. This parameter is an extension to urn:ietf:params:oauth:grant-type:uma-ticket grant type in order to allow clients to send authorization requests without a permission ticket. The format of the string must be: RESOURCE_ID#SCOPE_ID. For instance: Resource A#Scope A, Resource A#Scope A, Scope B, Scope C, Resource A, #Scope A.

So the question would be: Is there already any tests for this "extension" functionality?

**Comment by GeorgeBekh** [comment]
@pedroigor, added test

**Comment by pedroigor** [comment]
@GeorgeBekh, tks.

I found one last thing that we need to include in your PR. You'll need to include a change to invalidate the cache entry you are introducing with the new method.

The invalidation should be done here https://github.com/keycloak/keycloak/blob/80e5227bcdbea5111f6eb1424a2effa6f0aacf7a/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/authorization/StoreFactoryCacheManager.java#L140.

Basically, you'll need another parameter to the method `permissionTicketUpdated` to pass the resource name so that you're able to include your new cache entry in the list of invalidations.

If you did not manage to do it (I know this is complicated when working for the first time) I can send you the changes.

**Comment by pedroigor** [comment]
Actually, in addition to that change you'll need another one to invalidate the new entry when a resource is updated. I'll send you the changes.

**Comment by pedroigor** [comment]
@GeorgeBekh, could you please include these [changes](https://github.com/pedroigor/keycloak/commit/3957e37312a82e3d53ba58fc7f4ad9ed40b7fca9)  in your PR ?

After cherrying-pick, make sure PR is a single commit with your changes and mine. Thanks.

**Comment by GeorgeBekh** [comment]
@pedroigor, I've added your changes. Thanks for the help, I wouldn't figure out caching on my own

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@GeorgeBekh, Thank you.
